### PR TITLE
[[ Perf ]] Speed up unicode text processing

### PR DESCRIPTION
**Pull request by livecodeali**
This patch cuts the time of various unicode text processing functionality by about 25%. By implementing our own version of the grapheme cluster breaking algorithm, we are able to circumvent the copying penalty incurred by using the ICU break iterators. 

This also fixes a couple of inconsistencies which arose from incorrect fast paths used for codeunit <-> grapheme index mapping - in particular with variation selectors (bug 16717) and other codepoints that combine in more complex ways eg Hangul (bug 14589)

[Review on Reviewable](https://reviewable.io/reviews/livecode/livecode/3484)

### CODE REVIEW

File: libfoundation/src/foundation-unicode.cpp [hunk 1]
```diff
@@ -2407,3 +2407,97 @@ bool MCUnicodeWildcardMatch(const void *source_chars, uindex_t source_length, bo
...
+    // GB 8a: Do not break between regional indicator symbols.
+    if (t_left_gcb == U_GCB_REGIONAL_INDICATOR && t_right_gcb == U_GCB_REGIONAL_INDICATOR)
+        return false;
+    
+    // GB 9: Do not break before extending characters.
+    if (t_right_gcb == U_GCB_EXTEND)
+        return false;
+
+    // GB 9a: Do not break before SpacingMarks
+    if (t_right_gcb == U_GCB_SPACING_MARK)
```
Comment (runrevmark) [inline libfoundation/src/foundation-unicode.cpp:2491]: I take it the U_GCB_SPACING_MARK attribute is SpacingMark without the exceptions listed in UAX#29?

Comment (livecodeali) [inline libfoundation/src/foundation-unicode.cpp:2491]: Indeed it is - I inadvertently verified this when trying to use MYANMAR SIGN SHAN COUNCIL TONE-3 as my spacing mark example in the tests, not realising that it was in the list of exceptions.

**Comment by runrevmark** [comment]
Reviewed 14 of 16 files at r1.
Review status: 14 of 16 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

_[engine/src/exec-strings.cpp, line 1509 [r1]](https://reviewable.io:443/reviews/livecode/livecode/3484#-K8TFiEvYki49rqLntCs:-K8TFiEvYki49rqLntCt:1298992908) ([raw file](https://github.com/livecode/livecode/blob/f2958887cdeb38d83fffb84677b0d7872633e81d/engine/src/exec-strings.cpp#L1509)):_
This could be done much more efficiently as it just requires a grapheme boundary check at the start of the range and the end of the range. Indeed, it might be worth putting a 'IsGraphemeClusterBoundary(string, index) in libfoundation so it can take advantage of the flags.

---

_[libfoundation/include/foundation.h, line 2510 [r1]](https://reviewable.io:443/reviews/livecode/livecode/3484#-K8TGOZ0ZSCF2B1M8AYd:-K8TGOZ0ZSCF2B1M8AYe:882623587) ([raw file](https://github.com/livecode/livecode/blob/f2958887cdeb38d83fffb84677b0d7872633e81d/libfoundation/include/foundation.h#L2510)):_
Is there a reason this is a public API? String checking should be entirely internal to the MCStringRef implementation.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3484)_

**Comment by runrevmark** [comment]
Review status: 14 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

_[libfoundation/include/foundation.h, line 2094 [r1]](https://reviewable.io:443/reviews/livecode/livecode/3484#-K8TJV-05wkNHh_d6LTA:-K8TJV-05wkNHh_d6LTB:1975759904) ([raw file](https://github.com/livecode/livecode/blob/f2958887cdeb38d83fffb84677b0d7872633e81d/libfoundation/include/foundation.h#L2094)):_
I don't think this definition is correct... I thought trivial meant that codeunits and graphemes mapped 1:1? I think Trivial should mean it is a string of unicode chars which have exactly the same properties as native chars. This means that: each grapheme is a single codeunit, it contains no characters which fold to more than one character.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3484)_

**Comment by livecodeali** [comment]
_[libfoundation/include/foundation.h, line 2510 [r1]](https://reviewable.io:443/reviews/livecode/livecode/3484#-K8TGOZ0ZSCF2B1M8AYd:-K8TQhMJjcrFveifcPFI:982770061) ([raw file](https://github.com/livecode/livecode/blob/f2958887cdeb38d83fffb84677b0d7872633e81d/libfoundation/include/foundation.h#L2510)):_
I suppose it would be possible to put it into the MCStringIsTrivial check - the reason it is currently public is that it is used in ide.cpp to determine whether the string can be tokenized as a c-string.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3484)_

**Comment by livecodeali** [comment]
_[libfoundation/include/foundation.h, line 2094 [r1]](https://reviewable.io:443/reviews/livecode/livecode/3484#-K8TJV-05wkNHh_d6LTA:-K8TWtLiz0BeVaJVoJyO:1090316906) ([raw file](https://github.com/livecode/livecode/blob/f2958887cdeb38d83fffb84677b0d7872633e81d/libfoundation/include/foundation.h#L2094)):_
 Ah, do you mean the 'non-combining characters' part, yes that is slightly inaccurate. Will change.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3484)_

**Comment by runrevmark** [comment]
Reviewed 5 of 6 files at r2.
Review status: 16 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

_[libfoundation/src/foundation-chunk.cpp, line 1034 [r2]](https://reviewable.io:443/reviews/livecode/livecode/3484#-K8V2Q26I9I-wFyzbccp:-K8V2Q26I9I-wFyzbccq:-1933015528) ([raw file](https://github.com/livecode/livecode/blob/ab8f92d33508c6dcc927f10b4a52ac2425001a64/libfoundation/src/foundation-chunk.cpp#L1034)):_
Surely if a string 'canbenative' then it is trivial?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3484)_

**Comment by livecodeali** [comment]
_[libfoundation/src/foundation-chunk.cpp, line 1034 [r2]](https://reviewable.io:443/reviews/livecode/livecode/3484#-K8V2Q26I9I-wFyzbccp:-K8V8gz_AzTUZ5lq-LeO:1764542040) ([raw file](https://github.com/livecode/livecode/blob/ab8f92d33508c6dcc927f10b4a52ac2425001a64/libfoundation/src/foundation-chunk.cpp#L1034)):_
Yes.. I guess  I was just covering against the case that something failed to set all the appropriate flags. I'll check and make sure it's ok to remove

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3484)_

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 514d387

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 514d387de7a8123dadec3771a3d0759e73af5fb1

**Comment by livecode-vulcan** [comment]
😞 test failure 514d387de7a8123dadec3771a3d0759e73af5fb1
- try-community-linux-x86: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1639)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/1655)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1634)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1645)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/1656)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1658)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1663)

**Comment by peter-b** [comment]
Build fix is obviously correct.

@livecode-vulcan review ok a131317

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 5a3863e

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 5a3863e34aa367cc30e9eb510380692b652a22b8

**Comment by livecode-vulcan** [comment]
😎 test success 5a3863e34aa367cc30e9eb510380692b652a22b8
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1654)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1670)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1649)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1660)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1671)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1673)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1678)

**Pull request merged**